pkg/issuer/acme/dns/dns.go

```diff
@@ -177,12 +177,12 @@ func extractChallengeSolverConfig(ch *cmacme.Challenge) (*cmacme.ACMEChallengeSo
 // solverForChallenge returns a Solver for the given providerName.
 // The providerName is the name of an ACME DNS-01 challenge provider as
 // specified on the Issuer resource for the Solver.
-func (s *Solver) solverForChallenge(ctx context.Context, issuer v1.GenericIssuer, ch *cmacme.Challenge) (solver, *cmacme.ACMEChallengeSolverDNS01, error) {
+func (s *Solver) solverForChallenge(ctx context.Context, ch *cmacme.Challenge) (solver, *cmacme.ACMEChallengeSolverDNS01, error) {
 	log := logf.FromContext(ctx, "solverForChallenge")
 	dbg := log.V(logf.DebugLevel)
 
-	resourceNamespace := s.ResourceNamespace(issuer)
-	canUseAmbientCredentials := s.CanUseAmbientCredentials(issuer)
+	resourceNamespace := ch.Namespace
+	canUseAmbientCredentials := s.CanUseAmbientCredentialsFromRef(ch.Spec.IssuerRef)
```

Comment:
@inteon The `resourceNamespace` is the namespace where the Issuer or ClusterIssuer credentials are loaded from.
For challenges linked to a ClusterIssuer, the `resourceNamespace` should be the cluster scoped namespace (`cert-manager`).
I think this is the root cause of:
 * https://github.com/cert-manager/cert-manager/issues/7331

Reply:
Right, the `Challenge` lives in the same namespace as the `Certificate`, but the `Secret` lives in the `resourceNamespace`.
This is a bug indeed, strange that this was not being tested.

Reply:
I propose this bugfix: https://github.com/cert-manager/cert-manager/pull/7339